### PR TITLE
[IMP] cleanup of base_address_city & base_address_extension

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
Rely on country.address_view_id instead of hacking the view on the fly. The
address layout (street VS street_name street_number street_number2) depends now
on the country of the user, not on the installed localisation. (it's now
configurable per country)

Unified street format to "Chaussee de Namur 40 - Appt 12"; the
configurable ones where actually wrong by country.

Merged base_address_city into base_address_extension, to avoid creating bridge
modules for no reason. Uses city_id instead of city if the country of the user
defines it AND if enforce_cities is set on this country.

base_address_extended is not anymore required by any l10n module; you can or not
use it depending on your preferences. (if you don't use it, a regex is used to deduce street
number for l10n that needs it, like mexico)

Improved post_init script for large databases.

Split of street removed on res.company. (not required by any l10n)

### CODE REVIEW

File: addons/base_address_extended/__init__.py [hunk 1]
```diff
@@ -7,4 +7,10 @@
 
 def _update_street_format(cr, registry):
     env = api.Environment(cr, SUPERUSER_ID, {})
-    env['res.partner'].search([])._compute_street_data()
+    partners = env['res.partner'].search([])
+    if len(partners) < 20000:
+        partners._compute_street_data()
+    else:
+        # For large DB, a perfect data that crashes is worse than an unperfect one that does not crash.
+        # Note that EDIs allos to set the street_number in the street_name, so it's correct.
```
Comment (h4818) [inline addons/base_address_extended/__init__.py:15]: A small typo in the comment here -> allows

File: addons/base_address_extended/models/__init__.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from . import res_company
-from . import res_country
 from . import res_partner
+from . import res_city
+from . import res_country
```
Comment (h4818) [inline addons/base_address_extended/models/__init__.py:6]: why move res_country? unnecessary change to the diff. I also usually prefer alphabetical order

File: addons/base_address_extended/models/res_city.py [hunk 3]
```diff
@@ -25,7 +25,6 @@ def name_get(self):
     @api.model
     def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get_uid=None):
         args = list(args or [])
-        # optimize out the default criterion of ``ilike ''`` that matches everything
-        if not (name == '' and operator == 'ilike'):
-            args += ['|', (self._rec_name, operator, name), ('zipcode', operator, name)]
+        if name:
+            args += ['|', (self._rec_name, operator, name), ('zipcode', '=', name)]
```
Comment (h4818) [inline addons/base_address_extended/models/res_city.py:29]: you only wan't to find exact matches on zipcode?

File: addons/l10n_pe/data/res_country_data.xml [hunk 4]
```diff
@@ -7,24 +7,21 @@
...
                 <div class="o_address_format">
-                    <field name="parent_id" invisible="1"/>
-                    <field name="type" invisible="1"/>
-                    <field name="street_name" placeholder="Street Name..." attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}" class="oe_read_only"/>
-                    <field name="street" placeholder="Street" class="oe_edit_only"/>
-                    <field name="street2" placeholder="Street2" invisible="1"/>
-                    <div class="o_row">
-                        <label for="street_number" class="oe_edit_only"/>
-                        <field name="street_number" attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"/>
+                    <field name="street" placeholder="Street..." class="o_address_street oe_read_only"/>
```
Comment (h4818) [inline addons/l10n_pe/data/res_country_data.xml:10]: why does pe get this special treatment - can we not use the method you used for mx, co ie. set the view_id on the country

File: addons/base_address_extended/views/base_address_extended.xml [hunk 5]
```diff
@@ -1,90 +1,67 @@
...
-                    <label for="street_number" class="oe_edit_only"/>
-                    <span> </span>
-                    <field name="street_number"/>
-                    <label for="street_number2" class="oe_edit_only"/>
-                    <field name="street_number2"/>
-                </div>
-            </xpath>
-        </field>
-    </record>
+     <record id="base.co" model="res.country">
```
Comment (h4818) [inline addons/base_address_extended/views/base_address_extended.xml:63]: what about nl, de, hr? - they also had different address formats before
Can we not apply something similar for countries that require city_id - I think cn is one?

File: odoo/tools/misc.py [hunk 6]
```diff
@@ -1637,3 +1637,13 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
         message.encode(),
         hash_function,
     ).hexdigest()
+
```
Comment (h4818) [inline odoo/tools/misc.py:1640]: I noticed you removed a blank line on base/models/res_partner - any reason you have an extra one here?

**Review by h4818 — COMMENTED** [review]
Looks quite good - significantly more clean, just a few questions...
You can squash the commits too.

File: addons/base_address_extended/__init__.py [hunk 7]
```diff
@@ -7,4 +7,10 @@
 
 def _update_street_format(cr, registry):
     env = api.Environment(cr, SUPERUSER_ID, {})
-    env['res.partner'].search([])._compute_street_data()
+    partners = env['res.partner'].search([])
+    if len(partners) < 20000:
+        partners._compute_street_data()
+    else:
+        # For large DB, a perfect data that crashes is worse than an unperfect one that does not crash.
```
Comment (poma-odoo) [inline addons/base_address_extended/__init__.py:14]: Hi, for cases similar to this, adding an async worker (similar to [oca's queue_job](https://github.com/OCA/queue)) would solves a lot of problems.

File: addons/base_address_extended/models/__init__.py (hunk 2 above)
Comment (poma-odoo) [reply addons/base_address_extended/models/__init__.py:6]: Alphabetical order, the module is under heavy changes, I vote for it.

File: addons/base_address_extended/views/base_address_extended.xml (hunk 5 above)
Comment (h4818) [reply addons/base_address_extended/views/base_address_extended.xml:63]: nl seems only required for reports
cn_city depends on base_address_extended
check de, hr?

**Comment by poma-odoo** [comment]
I have a proposal based on @william-andre 's idea, what if we make address fields compute by default, so each l10n module can customize it ? I can explain better with code example:

In `odoo/odoo/addons/base/models/res_partner.py`:
```python
    street = fields.Char(compute='_compute_street', readonly=False, store=True)
    #...
    def _compute_street(self):
        pass
```
in `l10n_xx/models/res_partner.py` (or if it is something general it can be done in `base_address_extension`):
```python
    @api.depend('street_xx`)
    def _compute_street(self):
        self.super()._compute_street()
        country_xx=self.env.ref('base.xx')
        if self.env.user.company_id.country_id == country_xx:
            for partner in self.filtered(lambda rec:rec.country_id == country_xx):
                partner.street = self.l10n_xx_parse_address()
```
In almost all countries the local address rules are mandatory for companies operating in the country doing local business, it justifies the conditions in the compute method. It does not harm anything in no l10n module is installed and it is invisible for the rest of code , so it will not break anything.

File: addons/base_address_extended/models/__init__.py (hunk 2 above)
Comment (william-andre) [reply addons/base_address_extended/models/__init__.py:5]: How is that alphabetical order changes? The company has no formatting anymore in the Form view.

File: addons/base_address_extended/models/res_country.py [hunk 8]
```diff
@@ -4,15 +4,10 @@
...
 
-    street_format = fields.Text(
-        help="Format to use for streets belonging to this country.\n\n"
-             "You can use the python-style string pattern with all the fields of the street "
-             "(for example, use '%(street_name)s, %(street_number)s' if you want to display "
-             "the street name, followed by a comma and the house number)"
-             "\n%(street_name)s: the name of the street"
-             "\n%(street_number)s: the house number"
-             "\n%(street_number2)s: the door number",
-        default='%(street_number)s/%(street_number2)s %(street_name)s', required=True)
```
Comment (william-andre) [inline addons/base_address_extended/models/res_country.py:18]: > Unified street format to "Chaussee de Namur 40 - Appt 12"; the
configurable ones where actually wrong by country.

Wasn't it just missing implementations instead of being wrong?

Russia, Singapore, South Africa, South Corea have different formats for instance
https://bitboost.com/ref/international-address-formats.html

**Review by william-andre — COMMENTED** [review]
Merging the two modules is a great idea (I wouldn't mind putting it in `base` tbh), but the implementation seems a bit too simplified as we are losing features.

File: addons/base_address_extended/models/res_country.py (hunk 8 above)
Comment (h4818) [reply addons/base_address_extended/models/res_country.py:18]: @william-andre  But these countries don't need the street number to separated out in an EDI document - so the usual `street` field should suffice

File: addons/base_address_extended/__init__.py (hunk 7 above)
Comment (h4818) [reply addons/base_address_extended/__init__.py:14]: I like this idea

**Comment by h4818** [comment]
@jco-odoo @smetl Are we still waiting on odoo/enterprise#22951 before we can go ahead with these PR's (see enterprise and upgrade PRs too)
@william-andre do we still need street format? 
Maybe we should have a discussion and document which EDI's require a separate street number

File: addons/base_address_extended/tests/test_street_fields.py [hunk 9]
```diff
@@ -10,164 +10,31 @@ class TestStreetFields(TransactionCase):
     @classmethod
     def setUpClass(cls):
         super(TestStreetFields, cls).setUpClass()
```
Comment (smetl) [inline addons/base_address_extended/tests/test_street_fields.py:12]: empty setUpClass can be removed

File: addons/base_address_extended/views/base_address_extended.xml (hunk 5 above)
Comment (h4818) [reply addons/base_address_extended/views/base_address_extended.xml:63]: Here's the analysis...

|l10n|city_id|enforce_cities|requires split street, number|depends base_address_extended|street_split|address form|
|----|-------|--------------|-----------------------------|-----------------------------|------------|------------|
|co| &check; | &check; |  | &check; | | extended|
|mx| &check;| &check; | &check; | &check; | | &check;|
|cn_city| &check; | ~&cross;~ &check; | | &check; | | ~?~ extended|
|cl||||&cross;|||
|pe|&check;|&check;||&check;||&check;|
|nl|||&check;||&check;||
|hr|||||||
|de||||||

@jco-odoo can you please confirm we can remove the base_address_extended dependecy from `l10n_cl`. Also, `l10n_cn_city` seems to define `city_id`'s, but does not modify the form - I have changed that.

File: addons/base_address_extended/tests/test_street_fields.py (hunk 9 above)
Comment (h4818) [reply addons/base_address_extended/tests/test_street_fields.py]: done

File: addons/base_address_extended/views/base_address_extended.xml (hunk 5 above)
Comment (jco-odoo) [reply addons/base_address_extended/views/base_address_extended.xml]: ok for l10n_cl (also we can remove the dependency on base_address_city).  For l10n_cn_city, it stays a weird module to me as the only thing it does, is provide cities for cn, but why would not we do it with one module for all city hungry countries of the world?  But we have it, so we could keep it and the only thing we would need to do, which is already the case is set enforce_cities on the China country to true.  

Because it might be a lot easier to apply the city_id to all the addresses (as was done before with the inherit of fields_view_get before, but we just apply it through the view here).  If you send from the Netherlands to China and you have l10n_cn_city installed then you can also immediately choose the city_id instead of typing it manually (I hope you have the zip as city names are in Chinese) and the address would not appear differently if you look at it from one company or the other.

File: addons/base_address_extended/__init__.py (hunk 1 above)
Comment (h4818) [reply addons/base_address_extended/__init__.py]: fixed

File: addons/base_address_extended/models/res_city.py (hunk 3 above)
Comment (h4818) [reply addons/base_address_extended/models/res_city.py]: done

File: addons/base_address_extended/views/base_address_extended.xml (hunk 5 above)
Comment (h4818) [reply addons/base_address_extended/views/base_address_extended.xml]: done

File: addons/base_address_extended/models/res_city.py [hunk 10]
```diff
@@ -25,7 +25,6 @@ def name_get(self):
...
         return self._search(args, limit=limit, access_rights_uid=name_get_uid)
```
[the remark this answers is not in the document]
Comment (h4818) [reply addons/base_address_extended/models/res_city.py]: done

File: addons/base_address_extended/views/base_address_extended.xml [hunk 11]
```diff
@@ -1,90 +1,73 @@
...
         </field>
     </record>
 
-    <record id="view_partner_address_structured_form" model="ir.ui.view">
-        <field name="name">view_partner_address_structured_form</field>
+    <record id="address_street_extended_city_form" model="ir.ui.view">
+        <field name="name">partner.form.address.extended.city_id</field>
         <field name="model">res.partner</field>
-        <field name="inherit_id" ref="base.view_partner_address_form"/>
+        <field name="priority" eval="900"/>
```
Comment (jco-odoo) [inline addons/base_address_extended/views/base_address_extended.xml:49]: Maybe it is better to remove the priority as we have the inherit_id anyways.

File: addons/base_address_extended/views/base_address_extended.xml [hunk 12]
```diff
@@ -1,90 +1,73 @@
...
-                    <span> </span>
-                    <field name="street_number" attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"/>
-                    <label for="street_number2" class="oe_edit_only"/>
-                    <field name="street_number2" attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"/>
-                </div>
-            </xpath>
+            <field name="city" position="attributes">
+                <attribute name="attrs">
+                    {'invisible': [('country_enforce_cities', '=', True),
+                                    '|', ('city_id', '!=', False), ('city', 'in', ['',False])]}
```
Comment (jco-odoo) [inline addons/base_address_extended/views/base_address_extended.xml:55]:
```suggestion
                                   '|', ('city_id', '!=', False), ('city', 'in', ['',False])]}
```
Maybe there is an better indentation idea

File: odoo/addons/base/views/res_partner_views.xml [hunk 13]
```diff
@@ -288,9 +288,6 @@
                                 </kanban>
                                 <form string="Contact / Address">
                                     <sheet>
-                                        <!-- parent_id and type fields needed in attrs in base_address_city module which overwrites
-                                        _fields_view_get() of partner. It would be better to put those fields there but the web client
-                                        dosen't support when a field is displayed several times in the same view.-->
```
Comment (jco-odoo) [inline odoo/addons/base/views/res_partner_views.xml:293]: Normally having twice the same field in the view became supported.  (why add this comment?)

File: addons/base_address_extended/__init__.py (hunk 7 above)
Comment (h4818) [reply addons/base_address_extended/__init__.py:14]: For now this will work - I will look at a cron in a separate PR.
Neither this method nor cron will impact the database load time since New dbs won't have any partners.
The cron will just be 'nice to have'

File: addons/base_address_extended/views/base_address_extended.xml [hunk 14]
```diff
@@ -1,90 +1,73 @@
...
+                    <field name="city_id"
+                           placeholder="City"
+                           class="o_address_city"
+                           attrs="{'invisible': [('country_enforce_cities', '=', False)]}"
+                           context="{'default_country_id': country_id, 'default_state_id': state_id, 'default_zipcode': zip}"/>
+                    <field name="city"
+                           placeholder="City"
+                           class="o_address_city"
+                           attrs="{'invisible': [('country_enforce_cities', '=', True),
+                                                  '|', ('city_id', '!=', False), ('city', 'in', ['',False])]}"/>
```
Comment (jco-odoo) [inline addons/base_address_extended/views/base_address_extended.xml:30]: indent

File: addons/l10n_pe/data/res_country_data.xml [hunk 15]
```diff
@@ -7,24 +7,21 @@
...
+                    <div class="oe_edit_only o_row">
+                        <field name="street_name" placeholder="Street" style="flex: 3 1 auto"/>
                         <span> </span>
-                        <label for="street_number2" class="oe_edit_only"/>
-                        <field name="street_number2" attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"/>
+                        <field name="street_number" placeholder="House #" style="flex: 1 1 auto"/>
+                        <span> - </span>
+                        <field name="street_number2" placeholder="Door #" style="flex: 1 1 auto"/>
```
Comment (jco-odoo) [inline addons/l10n_pe/data/res_country_data.xml:16]: District and city is a bit uncertain for the moment, but we do not need to split the street anywhere, so we can simplify this view and simply remove it.

File: addons/l10n_pe/data/res_country_data.xml [hunk 16]
```diff
@@ -7,24 +7,21 @@
...
                     </div>
+                    <field name="street2" placeholder="Street 2..." class="o_address_street"/>
                     <field name="l10n_pe_district" placeholder="District..."/>
-                    <field name="city" placeholder="Province..." invisible="1"/>
-                    <field name="city_id" placeholder="Province..."/>
-                    <field name="state_id" class="o_address_state" placeholder="State..." options='{"no_open": True}'/>
-                    <field name="country_id" placeholder="Country" class="o_address_country" options='{"no_open": True, "no_create": True}'/>
+                    <field name="city_id" placeholder="Province..." class="o_address_city"/>
```
Comment (jco-odoo) [inline addons/l10n_pe/data/res_country_data.xml:20]: We should apply the same logic here with the enforce_cities as we have in the base_address_extended view. + It might be good to check how this view behaves further, because it has always been a bit dodgy.

File: odoo/addons/base/models/res_partner.py [hunk 17]
```diff
@@ -240,6 +239,10 @@ def default_get(self, default_fields):
         ('check_name', "CHECK( (type='contact' AND name IS NOT NULL) or (type!='contact') )", 'Contacts require a name'),
     ]
 
+    def _get_street_split(self):
```
Comment (jco-odoo) [inline odoo/addons/base/models/res_partner.py:242]: if we have it in tools, why put it here gain?

File: odoo/addons/base/models/res_partner.py [hunk 18]
```diff
@@ -27,7 +27,6 @@
                    ]
 WARNING_HELP = 'Selecting the "Warning" option will notify user with the message, Selecting "Blocking Message" will throw an exception with the message and block the flow. The Message has to be written in the next field.'
 
-
```
Comment (jco-odoo) [inline odoo/addons/base/models/res_partner.py:30]: unnecessary diff?

File: addons/base_address_extended/models/res_city.py [hunk 19]
```diff
@@ -25,7 +25,8 @@ def name_get(self):
...
         args = list(args or [])
-        # optimize out the default criterion of ``ilike ''`` that matches everything
-        if not (name == '' and operator == 'ilike'):
-            args += ['|', (self._rec_name, operator, name), ('zipcode', operator, name)]
-        return self._search(args, limit=limit, access_rights_uid=name_get_uid)
+        domain = []
+        if name:
+            connector = '!' if operator in expression.NEGATIVE_TERM_OPERATORS else '|'
+            domain = [connector, ('zipcode', 'ilike', name), (self._rec_name, operator, name)]
+        return self._search(expression.AND([domain, args]), limit=limit, access_rights_uid=name_get_uid)
```
Comment (smetl) [inline addons/base_address_extended/models/res_city.py:32]: plz comment this override to explain why you made it :)

File: addons/base_address_extended/models/res_country.py [hunk 20]
```diff
@@ -4,15 +4,10 @@
 from odoo import fields, models
 
 
-class ResCountry(models.Model):
+class Country(models.Model):
```
Comment (smetl) [inline addons/base_address_extended/models/res_country.py:7]: useless diff

File: addons/base_address_extended/models/res_partner.py [hunk 21]
```diff
@@ -17,123 +13,39 @@ class Partner(models.Model):
     street_number2 = fields.Char(
         'Door', compute='_compute_street_data', inverse='_inverse_street_data', store=True)
 
+    city_id = fields.Many2one('res.city', string='City')
```
Comment (smetl) [inline addons/base_address_extended/models/res_partner.py]: plz specify the `comodel_name` keyword

File: addons/base_address_extended/models/res_partner.py [hunk 22]
```diff
@@ -17,123 +13,39 @@ class Partner(models.Model):
...
+    country_enforce_cities = fields.Boolean(related='country_id.enforce_cities', readonly=True)
```
Comment (smetl) [inline addons/base_address_extended/models/res_partner.py:17]: by default, a `related` field is `readonly` so you can drop the last parameter

File: addons/base_address_extended/models/res_partner.py [hunk 23]
```diff
@@ -17,123 +13,39 @@ class Partner(models.Model):
...
-                previous_field = field_name
-                previous_pos = re_match.end()
-
-            # add trailing chars in street_format
-            street_value += street_format[previous_pos:]
-            partner.street = street_value
+            street = ((partner.street_name or '') + " " + (partner.street_number or '')).strip()
+            if partner.street_number2:
+                street = street + " - " + partner.street_number2
+            partner.street = street
```
Comment (smetl) [inline addons/base_address_extended/models/res_partner.py:25]: In my opinion, this is making this module completely useless... There is different street formats all around the world. We can't just drop it like this. Do we plan to add manually all street formatting later like
```py
if partner.country_code == 'IN':
    street_format = ...
elif ...
```
?

File: addons/base_address_extended/models/res_partner.py [hunk 24]
```diff
@@ -17,123 +13,39 @@ class Partner(models.Model):
...
+    @api.onchange('city_id')
+    def _onchange_city_id(self):
+        if self.city_id:
+            self.city = self.city_id.name
+            self.zip = self.city_id.zipcode
+            self.state_id = self.city_id.state_id
+        elif self._origin:
+            self.city = False
+            self.zip = False
+            self.state_id = False
```
Comment (smetl) [inline addons/base_address_extended/models/res_partner.py:51]: I don't understand why `False` in case of `_origin`

File: addons/base_address_extended/views/base_address_extended.xml [hunk 25]
```diff
@@ -1,90 +1,72 @@
...
-                    <span> </span>
-                    <field name="street_number" attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"/>
-                    <label for="street_number2" class="oe_edit_only"/>
-                    <field name="street_number2" attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"/>
-                </div>
-            </xpath>
+            <field name="city" position="attributes">
+                <attribute name="attrs">
+                    {'invisible': [('country_enforce_cities', '=', True),
+                                    '|', ('city_id', '!=', False), ('city', 'in', ['',False])]}
```
Comment (smetl) [inline addons/base_address_extended/views/base_address_extended.xml:54]: `<attribute name="attrs">{'invisible': [('country_enforce_cities', '=', True), '|', ('city_id', '!=', False), ('city', 'in', ('',False))]}</attribute>`

File: addons/base_address_extended/models/res_country.py [hunk 26]
```diff
@@ -4,15 +4,10 @@
...
-             "(for example, use '%(street_name)s, %(street_number)s' if you want to display "
-             "the street name, followed by a comma and the house number)"
-             "\n%(street_name)s: the name of the street"
-             "\n%(street_number)s: the house number"
-             "\n%(street_number2)s: the door number",
-        default='%(street_number)s/%(street_number2)s %(street_name)s', required=True)
+    enforce_cities = fields.Boolean(
+        string='Enforce Cities',
+        help="Check this box to ensure every address created in that country has a 'City' chosen "
+             "in the list of the country's cities.")
```
Comment (smetl) [inline addons/base_address_extended/models/res_country.py:13]: Do we still need this checkbox? We could add `city_id` into the address view using the custom `ir.ui.view` when needed

File: addons/base_address_extended/models/res_partner.py (hunk 24 above)
Comment (poma-odoo) [reply addons/base_address_extended/models/res_partner.py:51]: it is  from original base_address_city:
https://github.com/odoo/odoo/commit/d7af73dcb1b4ef3ada39bcafa7ad23972e487b45

File: addons/base_address_extended/models/res_partner.py [hunk 27]
```diff
@@ -17,123 +13,39 @@ class Partner(models.Model):
     street_number2 = fields.Char(
         'Door', compute='_compute_street_data', inverse='_inverse_street_data', store=True)
 
+    city_id = fields.Many2one('res.city', string='City ID')
```
[the remark this answers is not in the document]
Comment (h4818) [reply addons/base_address_extended/models/res_partner.py:16]: we can't have the same description - it raises warnings. Since the label is not visible, this is ok

File: odoo/addons/base/models/res_partner.py (hunk 18 above)
Comment (h4818) [reply odoo/addons/base/models/res_partner.py]: done

File: addons/l10n_pe/data/res_country_data.xml (hunk 16 above)
Comment (h4818) [reply addons/l10n_pe/data/res_country_data.xml]: done

File: addons/l10n_pe/data/res_country_data.xml (hunk 15 above)
Comment (h4818) [reply addons/l10n_pe/data/res_country_data.xml]: done

File: addons/base_address_extended/models/res_partner.py [hunk 28]
```diff
@@ -17,123 +13,39 @@ class Partner(models.Model):
...
+    country_enforce_cities = fields.Boolean(related='country_id.enforce_cities', readonly=True)
+
     def _inverse_street_data(self):
-        """Updates the street field.
-        Writes the `street` field on the partners when one of the sub-fields in STREET_FIELDS
-        has been touched"""
-        street_fields = self._get_street_fields()
+        """ update self.street based on street_name, street_nuber and street_number2 """
```
[the remark this answers is not in the document]
Comment (h4818) [reply addons/base_address_extended/models/res_partner.py]: done

File: addons/base_address_extended/models/res_partner.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (h4818) [reply addons/base_address_extended/models/res_partner.py:16]: it's not visible

File: addons/base_address_extended/models/res_country.py (hunk 26 above)
Comment (h4818) [reply addons/base_address_extended/models/res_country.py:13]: Indeed, with base_address_extended we update the default form view to use city_id if enforce_countries. This allows to enable city_id for any country

File: addons/base_address_extended/__manifest__.py [hunk 29]
```diff
@@ -13,10 +14,14 @@
 
 For example, in legal reports, some countries need to split the street into several fields,
 with the street name, the house number, and room number.
+
+Use this module only when city_id is required. Standard street_split tools are available otherwise.
```
[the remark this answers is not in the document]
Comment (h4818) [reply addons/base_address_extended/__manifest__.py]: changed

File: addons/base_address_extended/views/base_address_extended.xml (hunk 12 above)
Comment (h4818) [reply addons/base_address_extended/views/base_address_extended.xml]: inlined instead...done

File: odoo/addons/base/views/res_partner_views.xml (hunk 13 above)
Comment (h4818) [reply odoo/addons/base/views/res_partner_views.xml:293]: removed, done

File: addons/base_address_extended/views/base_address_extended.xml (hunk 14 above)
Comment (h4818) [reply addons/base_address_extended/views/base_address_extended.xml]: inlined, done

File: odoo/addons/base/models/res_partner.py (hunk 17 above)
Comment (h4818) [reply odoo/addons/base/models/res_partner.py:243]: tools is generic for and street argument
_get_street_split(): calls tools.street_split for the partner street
_get_street_split() base_address_extended override uses the stored fields first

File: addons/base_address_extended/__init__.py (hunk 7 above)
Comment (h4818) [reply addons/base_address_extended/__init__.py:14]: @smetl you were right - `compute` and `inverse` are both called, however theres something more interesting...
On module installation, after creating the database tables, the stored computed fields are prepared ... so `_compute_street_data` is executed once already.
*After* that, the post_init_hook is called and `_update_street_format` does this for a second time. And this time it executes both compute and inverse...

So in an attempt to gain performance, we are actually wasting it :)

Comment (h4818) [reply addons/base_address_extended/__init__.py]: removed the post_init_hook since it is doing a computation that is already done by the ORM for stored, computed fields

File: addons/base_address_extended/models/res_city.py (hunk 19 above)
Comment (h4818) [reply addons/base_address_extended/models/res_city.py:33]: done

File: addons/base_address_extended/models/res_country.py (hunk 20 above)
Comment (h4818) [reply addons/base_address_extended/models/res_country.py:7]: it's just for consistency and uniformity

File: addons/base_address_extended/models/res_partner.py (hunk 24 above)
Comment (h4818) [reply addons/base_address_extended/models/res_partner.py:51]: basically, what @poma-odoo said, and...
if the city_id is set, we copy the city_id field values to the partner fields,
but if you clear the city_id, only reset the partner fields if it's already in the db.

File: addons/base_address_extended/views/base_address_extended.xml (hunk 25 above)
Comment (h4818) [reply addons/base_address_extended/views/base_address_extended.xml]: inlined

File: addons/base_address_extended/models/res_partner.py (hunk 22 above)
Comment (h4818) [reply addons/base_address_extended/models/res_partner.py]: done

File: addons/l10n_co/__manifest__.py [hunk 30]
```diff
@@ -17,7 +17,7 @@
     'website': 'https://www.odoo.com/colombia',
     'depends': [
         'account',
-        'base_address_city',
+        'base_address_extended',
```
Comment (jco-odoo) [inline addons/l10n_co/__manifest__.py:20]: This dependency is weird, because the cities are loaded in l10n_co_edi (because of the extra field on them)

Comment (h4818) [reply addons/l10n_co/__manifest__.py]: moved to l10n_co_edi

**Review by smetl — APPROVED** [review]

**Comment by smetl** [comment]
@robodoo r+
